### DISCUSSION TITLE
CREATION OF MultiFab WITH OBJECTS AT EACH COMPONENT

### DISCUSSION THREAD

**Discussion by FernandoIGuerra** [AMReX-Combustion/PeleC#818] [Q&A]
Hi, Everyone,
Is it possible to allocate at each component of the MultiFab and object. So far, I implemented the code shown below. However, the compiler indicates that there is an conversion error from Array4<double> to non-scalar Array4<LEMElement<double>>  The intuitive idea behind this is ti create LEM elements inside each LES cell. Thanks in advance.
```c++
template<typename T>
class LEMElement {
private:
    T* primitive_variables;
    T* position;
    size_t primitive_variables_size;
    size_t position_size;

amrex::MultiFab lem_mf(ba, dm, num_lem_elements, 0);

        for(amrex::MFIter mfi(lem_mf); mfi.isValid(); ++mfi){
                	const amrex::Box& bx = mfi.validbox();
                	const amrex::Array4<LEMElement<double>>mf_array = lem_mf.array(mfi);
                	amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k){
                		mf_array(i,j,k, 0) = LEMElement<double>(primitive_size, position_size);
                	});
                }
```

**Comment by marchdf** [comment]
AFAIK, MultiFabs cannot contain arbitrary objects. They are defined for floats/doubles and ints (`iMultiFab`).  You would want to use `DeviceVector` to hold arbitrary objects. Then you can map from the cell index i/j/k to a vector index: box.index(iv) or something to that effect. Things might get messy but it's been done before.

**Reply by drummerdoc** [reply]
Depending on exactly how you want to implement things, you can make a `FabArray<FAB>` where FAB is `BaseFab<T>` and T is the type of data at every index.   However, it will take a bit of work to get something nicely usable within the ParallelFor.  For example it might be that T has to be POD (plain old data) and therefore easily converted to a C-array of data stacked up in a standard way that Array4 indexing can access what you expect.

**Reply by drummerdoc** [reply]
PS: I suggest you take this question to the AMReX issues board, as there will be many folks there with suggestions!

**Reply by FernandoIGuerra** [reply]
Another option is to create another mesh with different BoxArray and communicate both meshes. My doubt is ParallerFor support the iteration of two BoxArrays at the same time.
